pkg/build/controller/controller.go

```diff
@@ -79,21 +81,74 @@ func (bc *BuildController) CancelBuild(build *buildapi.Build) error {
 // HandleBuild deletes pods for cancelled builds and takes new builds and puts
 // them in the pending state after creating a corresponding pod
 func (bc *BuildController) HandleBuild(build *buildapi.Build) error {
-	glog.V(4).Infof("Handling build %s/%s", build.Namespace, build.Name)
+	glog.V(4).Infof("Handling build %s/%s (%s)", build.Namespace, build.Name, build.Status.Phase)
+
+	// If we receive build that reached its final state and the build was created
+	// from a serial build configuration, tag the next build in the queue for
+	// execution. That will produce event and the build should get executed
+	// immediately.
+	var nextBuild *buildapi.Build
+	schedulePolicy := buildutil.BuildSchedulePolicy(build)
+	if schedulePolicy == buildapi.BuildSchedulePolicySerial {
+		var err error
+		bcName := buildutil.BuildConfigName(build)
+		if len(bcName) == 0 {
+			return fmt.Errorf("build %s/%s has no build config label set", build.Namespace, build.Name)
+		}
+		if nextBuild, err = bc.nextBuild(build.Namespace, bcName); err != nil {
+			return err
+		}
+		if nextBuild != nil {
+			glog.V(5).Infof("Build configuration %s/%s is serial, next build is %s", build.Namespace, bcName, nextBuild.Name)
+			if buildutil.IsBuildComplete(build) {
+				bc.Recorder.Eventf(nextBuild, kapi.EventTypeNormal, "scheduled", "Scheduling build %s/%s to run next", nextBuild.Namespace, nextBuild.Name)
+				now := unversioned.Now()
+				nextBuild.Status.StartTimestamp = &now
+				err := bc.BuildUpdater.Update(nextBuild.Namespace, nextBuild)
+				if errors.IsConflict(err) {
+					// If the build was updated, it is most likely because it was
+					// cancelled. In that case, wait for the retry controller to kick the
+					// next build, which might take longer, but it is better than guessing
+					// the next build.
+					return nil
+				}
+				return err
+			}
+		}
+	}
 
 	// A cancelling event was triggered for the build, delete its pod and update build status.
-	if build.Status.Cancelled && build.Status.Phase != buildapi.BuildPhaseCancelled {
-		if err := bc.CancelBuild(build); err != nil {
-			build.Status.Reason = buildapi.StatusReasonCancelBuildFailed
-			return fmt.Errorf("Failed to cancel build %s/%s: %v, will retry", build.Namespace, build.Name, err)
+	if build.Status.Phase != buildapi.BuildPhaseCancelled {
+		var reason buildapi.StatusReason
+		switch {
```

Comment:
@kargakis I kind of like switch because it makes it better to read ;-) is there any pragmatic reason to use if here?